db/blob_file_state.h

```diff
@@ -0,0 +1,93 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+#include <iosfwd>
+#include <string>
+
+namespace rocksdb {
+
+constexpr uint64_t kInvalidBlobFileNumber = 0;
+
+class JSONWriter;
+class Slice;
+class Status;
+
+class BlobFileState {
+ public:
+  BlobFileState() = default;
+
+  BlobFileState(uint64_t blob_file_number, uint64_t total_blob_count,
+                uint64_t total_blob_bytes)
+      : blob_file_number_(blob_file_number),
+        total_blob_count_(total_blob_count),
+        total_blob_bytes_(total_blob_bytes) {}
+
+  BlobFileState(uint64_t blob_file_number, uint64_t total_blob_count,
+                uint64_t total_blob_bytes, uint64_t garbage_blob_count,
+                uint64_t garbage_blob_bytes)
+      : blob_file_number_(blob_file_number),
+        total_blob_count_(total_blob_count),
+        total_blob_bytes_(total_blob_bytes),
+        garbage_blob_count_(garbage_blob_count),
+        garbage_blob_bytes_(garbage_blob_bytes) {
+    assert(garbage_blob_count_ <= total_blob_count_);
+    assert(garbage_blob_bytes_ <= total_blob_bytes_);
+  }
+
+  uint64_t GetBlobFileNumber() const { return blob_file_number_; }
+
+  uint64_t GetTotalBlobCount() const { return total_blob_count_; }
+  uint64_t GetTotalBlobBytes() const { return total_blob_bytes_; }
+
+  void AddGarbageBlob(uint64_t size) {
+    assert(garbage_blob_count_ < total_blob_count_);
+    assert(garbage_blob_bytes_ + size <= total_blob_bytes_);
+
+    ++garbage_blob_count_;
+    garbage_blob_bytes_ += size;
+  }
+
+  uint64_t GetGarbageBlobCount() const { return garbage_blob_count_; }
+  uint64_t GetGarbageBlobBytes() const { return garbage_blob_bytes_; }
+
+  bool IsObsolete() const {
+    assert(garbage_blob_count_ <= total_blob_count_);
+
+    return !(garbage_blob_count_ < total_blob_count_);
+  }
+
+  void EncodeTo(std::string* output) const;
+  Status DecodeFrom(Slice* input);
+
+  std::string DebugString() const;
+  std::string DebugJSON() const;
+
+  bool operator==(const BlobFileState& rhs) const {
```

Comment:
We follow Google C++ style guide, which discourages overloading `==` as member function. https://google.github.io/styleguide/cppguide.html#Operator_Overloading

Reply:
That only makes a difference if implicit conversions are in play (which is not the case here) but will change it to be in sync. Thanks!